core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/BedrockInventoryTransactionTranslator.java

```diff
@@ -537,8 +540,8 @@ private boolean isIncorrectHeldItem(GeyserSession session, InventoryTransactionP
     }
 
     private boolean useItem(GeyserSession session, InventoryTransactionPacket packet, int blockState) {
-        // Let the server decide if the item should change, not the client, and revert the changes the client made
-        InventoryTranslator.PLAYER_INVENTORY_TRANSLATOR.updateSlot(session, session.getPlayerInventory(), session.getPlayerInventory().getOffsetForHotbar(packet.getHotbarSlot()));
+        // Update the player's inventory to remove any items added by the client itself
+        InventoryTranslator.PLAYER_INVENTORY_TRANSLATOR.updateInventory(session, session.getPlayerInventory());
```

Comment:
This will resend the entire inventory contents every time this is called. I'd rather not; why did you change this?

Reply:
If you have a stack of glass bottles or buckets and click above the water on an adjacent block, the client adds a water bottle/water bucket to the next empty slot in the player's inventory.
I guess it would be better to just update the empty slots in the inventory?

Reply:
Any chance we can replicate that logic on our end and revert it?

Reply:
I'll take a look.